src/main/cpp/src/case_when.cu

```diff
@@ -0,0 +1,151 @@
+  cudf::table_device_view const d_table;
+
+  /**
+   * The number of bool columns is the size of case when branches.
+   * Note: reuturned index may be out of bound, valid bound is [0, col_num)
+   * When returning col_num index, it means final result is NULL value or ELSE value.
+   *
+   * e.g.:
+   * CASE WHEN 'a' THEN 'A' END
+   *   The number of bool columns is 1
+   *   The number of scalars is 1
+   *   Max index is 1 which means using NULL(all when exprs are false).
+   * CASE WHEN 'a' THEN 'A' ELSE '_' END
+   *   The number of bool columns is 1
+   *   The number of scalars is 2
+   *   Max index is also 1 which means using else value '_'
+   */
+  __device__ cudf::size_type operator()(std::size_t row_idx)
+  {
+    auto col_num                     = d_table.num_columns();
+    bool found_true                  = false;
+    cudf::size_type first_true_index = col_num;
+    for (auto col_idx = 0; !found_true && col_idx < col_num; col_idx++) {
+      auto const& col = d_table.column(col_idx);
+      if (!col.is_null(row_idx) && col.element<bool>(row_idx)) {
+        // Predicate is true and not null
+        found_true       = true;
+        first_true_index = col_idx;
+      }
+    }
+    return first_true_index;
+  }
+};
+
+std::unique_ptr<cudf::column> select_first_true_index(cudf::table_view const& when_bool_columns,
+                                                      rmm::cuda_stream_view stream,
+                                                      rmm::mr::device_memory_resource* mr)
+{
+  // checks
+  auto const num_columns = when_bool_columns.num_columns();
+  CUDF_EXPECTS(num_columns > 0, "At least one column must be specified");
+  auto const row_count = when_bool_columns.num_rows();
+  if (row_count == 0)  // empty begets empty
+    return cudf::make_empty_column(cudf::type_id::INT32);
+
+  // make output column
+  auto ret = cudf::make_numeric_column(
+    cudf::data_type{cudf::type_id::INT32}, row_count, cudf::mask_state::ALL_VALID, stream, mr);
+
+  // select first true index
+  auto d_table = cudf::table_device_view::create(when_bool_columns, stream);
+  thrust::transform(rmm::exec_policy(stream),
+                    thrust::make_counting_iterator<cudf::size_type>(0),
+                    thrust::make_counting_iterator<cudf::size_type>(row_count),
+                    ret->mutable_view().begin<cudf::size_type>(),
+                    select_first_true_fn{*d_table});
+  return ret;
+}
+
+std::unique_ptr<cudf::column> select_from_index(
+  cudf::strings_column_view const& then_and_else_scalar_column,
+  cudf::column_view const& select_index_column,
+  rmm::cuda_stream_view stream,
+  rmm::mr::device_memory_resource* mr)
+{
+  cudf::size_type num_of_rows   = select_index_column.size();
+  cudf::size_type num_of_scalar = then_and_else_scalar_column.size();
+
+  // create device views
+  auto d_scalars =
+    *(cudf::column_device_view::create(then_and_else_scalar_column.parent(), stream));
```

Comment:
Oh never use this again. This is dereference of a temporary object which is destroyed immediately.

Reply:
It is no longer applicable. Because `select_from_index` was removed.